backend/main.go

```
@@ -58,22 +75,52 @@ func assignDropRates() {
 		dropRates[7] = [5]float64{0.15, 0.20, 0.35, 0.25, 0.05}
 		dropRates[8] = [5]float64{0.10, 0.15, 0.30, 0.30, 0.15}
 
-		for i, s := range dropRates {
+		file, err := json.Marshal(dropRates)
+		if err != nil {
+			fmt.Printf("Error happened in JSON marshal. Err: %s", err)
+			return errors.New("could not retrieve droprates")
+		}
+		_ = ioutil.WriteFile("./data/dropRates.json", file, 0644)
+	}
+
+	// convert drop rates to cumulative percentages
+	dropRatesConvOK := true
+	cFile, cErr := os.Open("./data/dropRatesConv.json")
+
+	if cErr == nil {
+		bv, err := ioutil.ReadAll(cFile)
+		if err != nil {
+			dropRatesConvOK = false
+		} else {
+			json.Unmarshal(bv, &dropRatesConv)
+		}
+	} else {
+		dropRatesConvOK = false
+	}
+	defer cFile.Close()
+
+	if !dropRatesConvOK {
+		dropRatesConv = dropRates
+		for i, s := range dropRatesConv {
 			for i2 := range s {
 				if i2 != 0 {
-					dropRates[i][i2] = math.Round((dropRates[i][i2]+dropRates[i][i2-1])*100) / 100
+					dropRatesConv[i][i2] = math.Round((dropRatesConv[i][i2]+dropRatesConv[i][i2-1])*100) / 100
 				}
 			}
 		}
 
-		file, _ := json.Marshal(dropRates)
-		_ = ioutil.WriteFile("./data/dropRates.json", file, 0644)
+		cFile, err := json.Marshal(dropRatesConv)
+		if err != nil {
+			fmt.Printf("Error happened in JSON marshal. Err: %s", err)
+			return errors.New("could not retrieve droprates")
```

You could just return the original error, perhaps adding additional information with `fmt.Errorf` if you'd like:

```
if err != nil {
  return fmt.Errorf("error happened in JSON Marshal: %w", err)
}
```

`%w` "wraps" the original error. This creates a new error with the given error message and allows callers to unwrap to get the original error with `errors.Unwrap(error)`

Try to limit logging statements in the middle of re-usable function calls

---

Don't think I ended up needing this. But will keep in mind.